src/dp_abstraction/aca_goal_state_handler.cpp

```diff
@@ -287,33 +316,39 @@ int Aca_Goal_State_Handler::update_router_states(GoalState &parsed_struct,
   int rc;
   int overall_rc = EXIT_SUCCESS;
   int count = 1;
+  GoalState* gs_ptr = &parsed_struct;
+  GoalStateOperationReply* reply_ptr = &gsOperationReply;
+  marl::WaitGroup wait_group(parsed_struct.router_states_size());
 
   for (int i = 0; i < parsed_struct.router_states_size(); i++) {
     ACA_LOG_DEBUG("=====>parsing router states #%d\n", i);
 
     RouterState current_RouterState = parsed_struct.router_states(i);
-
-    workitem_future.push_back(std::async(
-            std::launch::async, &Aca_Goal_State_Handler::update_router_state_workitem, this,
-            current_RouterState, std::ref(parsed_struct), std::ref(gsOperationReply)));
-    if (count % resource_state_processing_batch_size == 0) {
-      for (int i = 0; i < workitem_future.size(); i++) {
-        rc = workitem_future[i].get();
-        if (rc != EXIT_SUCCESS)
-          overall_rc = rc;
-      }
-      workitem_future.clear();
-      count = 1;
-    } else {
-      count ++;
-    }
-  }
-
-  for (int i = 0; i < workitem_future.size(); i++) {
-    rc = workitem_future[i].get();
-    if (rc != EXIT_SUCCESS)
-      overall_rc = rc;
+    marl::schedule([=] {
+      defer(wait_group.done());
+      update_router_state_workitem(current_RouterState, *gs_ptr, *reply_ptr);
+    });
+    // workitem_future.push_back(std::async(
+    //         std::launch::async, &Aca_Goal_State_Handler::update_router_state_workitem, this,
+    //         current_RouterState, std::ref(parsed_struct), std::ref(gsOperationReply)));
+    // if (count % resource_state_processing_batch_size == 0) {
+    //   for (int i = 0; i < workitem_future.size(); i++) {
+    //     rc = workitem_future[i].get();
+    //     if (rc != EXIT_SUCCESS)
+    //       overall_rc = rc;
+    //   }
+    //   workitem_future.clear();
+    //   count = 1;
+    // } else {
+    //   count ++;
+    // }
   }
+  wait_group.wait();
```

Comment:
remove unstable wait_group

Reply:
We shall keep the waitgroup as we need to wait for all states to be finished until going into the next step.